File: public/docs/sdks/go/smart_cities.go
package lydian

import (
	"context"
	"encoding/json"
	"fmt"
)

// SmartCitiesClient handles Smart Cities API operations
type SmartCitiesClient struct {
	client *Client
}

// City represents a city entity
type City struct {
	ID         string                 `json:"id"`
	Name       string                 `json:"name"`
	Country    string                 `json:"country"`
	Population *int                   `json:"population,omitempty"`
	Timezone   string                 `json:"timezone,omitempty"`
	Metadata   map[string]interface{} `json:"metadata,omitempty"`
	CreatedAt  string                 `json:"created_at,omitempty"`
	UpdatedAt  string                 `json:"updated_at,omitempty"`
}

// CityAsset represents a city asset
type CityAsset struct {
	ID        string                 `json:"id"`
	CityID    string                 `json:"city_id"`
	Type      string                 `json:"type"`
	Name      string                 `json:"name"`
	Location  Location               `json:"location"`
	Status    string                 `json:"status"`
	Metadata  map[string]interface{} `json:"metadata,omitempty"`
	CreatedAt string                 `json:"created_at,omitempty"`
	UpdatedAt string                 `json:"updated_at,omitempty"`
}

// Location represents geographic coordinates
type Location struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
	Address   string  `json:"address,omitempty"`
}

// Alert represents a city alert
type Alert struct {
	ID          string    `json:"id"`
	CityID      string    `json:"city_id"`
	Type        string    `json:"type"`
	Severity    string    `json:"severity"`
	Title       string    `json:"title"`
	Description string    `json:"description"`
	Location    *Location `json:"location,omitempty"`
	Status      string    `json:"status"`
	CreatedAt   string    `json:"created_at,omitempty"`
	ResolvedAt  *string   `json:"resolved_at,omitempty"`
}

// CreateCity creates a new city
func (sc *SmartCitiesClient) CreateCity(ctx context.Context, city City) (*City, error) {
	data, err := sc.client.Request(ctx, "POST", "/smart-cities/cities", city, nil)
	if err != nil {
		return nil, err
	}

	var result City
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}

	return &result, nil
}

// GetCity retrieves a city by ID
func (sc *SmartCitiesClient) GetCity(ctx context.Context, cityID string) (*City, error) {
	data, err := sc.client.Request(ctx, "GET", fmt.Sprintf("/smart-cities/cities/%s", cityID), nil, nil)
	if err != nil {
		return nil, err
	}

	var city City
	if err := json.Unmarshal(data, &city); err != nil {
		return nil, err
	}

	return &city, nil
}

// ListCities lists all cities
func (sc *SmartCitiesClient) ListCities(ctx context.Context, params PaginationParams) (*PaginatedResponse, error) {
	query := make(map[string]string)
	if params.Page > 0 {
		query["page"] = fmt.Sprintf("%d", params.Page)
	}
	if params.Limit > 0 {
		query["limit"] = fmt.Sprintf("%d", params.Limit)
	}
	if params.Cursor != "" {
		query["cursor"] = params.Cursor
	}

	data, err := sc.client.Request(ctx, "GET", "/smart-cities/cities", nil, query)
	if err != nil {
		return nil, err
	}

	var result PaginatedResponse
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}

	return &result, nil
}

// CreateAsset creates a city asset
func (sc *SmartCitiesClient) CreateAsset(ctx context.Context, asset CityAsset) (*CityAsset, error) {
	data, err := sc.client.Request(ctx, "POST", "/smart-cities/assets", asset, nil)
	if err != nil {
		return nil, err
	}

	var result CityAsset
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}

	return &result, nil
}

// CreateAlert creates a city alert
func (sc *SmartCitiesClient) CreateAlert(ctx context.Context, alert Alert) (*Alert, error) {
	data, err := sc.client.Request(ctx, "POST", "/smart-cities/alerts", alert, nil)
	if err != nil {
		return nil, err
	}

	var result Alert
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}

	return &result, nil
}

File: public/docs/sdks/typescript/src/smart-cities.ts
import { LydianClient } from './client';
import { City, CityAsset, CityMetrics, Alert, PaginatedResponse, PaginationParams } from './types';

/**
 * Smart Cities API client
 */
export class SmartCitiesClient {
  constructor(private client: LydianClient) {}

  /**
   * Create a new city
   */
  async createCity(data: Omit<City, 'id' | 'createdAt' | 'updatedAt'>): Promise<City> {
    return this.client.request<City>('POST', '/smart-cities/cities', { body: data });
  }

  /**
   * Get city by ID
   */
  async getCity(cityId: string): Promise<City> {
    return this.client.request<City>('GET', `/smart-cities/cities/${cityId}`);
  }

  /**
   * List all cities with pagination
   */
  async listCities(params?: PaginationParams): Promise<PaginatedResponse<City>> {
    return this.client.request<PaginatedResponse<City>>('GET', '/smart-cities/cities', {
      query: params,
    });
  }

  /**
   * Update city
   */
  async updateCity(cityId: string, data: Partial<City>): Promise<City> {
    return this.client.request<City>('PATCH', `/smart-cities/cities/${cityId}`, { body: data });
  }

  /**
   * Delete city
   */
  async deleteCity(cityId: string): Promise<void> {
    await this.client.request<void>('DELETE', `/smart-cities/cities/${cityId}`);
  }

  /**
   * Create city asset
   */
  async createAsset(data: Omit<CityAsset, 'id' | 'createdAt' | 'updatedAt'>): Promise<CityAsset> {
    return this.client.request<CityAsset>('POST', '/smart-cities/assets', { body: data });
  }

  /**
   * List city assets
   */
  async listAssets(cityId: string, params?: PaginationParams): Promise<PaginatedResponse<CityAsset>> {
    return this.client.request<PaginatedResponse<CityAsset>>('GET', `/smart-cities/cities/${cityId}/assets`, {
      query: params,
    });
  }

  /**
   * Get city metrics
   */
  async getMetrics(cityId: string, startDate?: string, endDate?: string): Promise<CityMetrics[]> {
    return this.client.request<CityMetrics[]>('GET', `/smart-cities/cities/${cityId}/metrics`, {
      query: { startDate, endDate },
    });
  }

  /**
   * Create alert
   */
  async createAlert(data: Omit<Alert, 'id' | 'createdAt'>): Promise<Alert> {
    return this.client.request<Alert>('POST', '/smart-cities/alerts', { body: data });
  }

  /**
   * List alerts
   */
  async listAlerts(cityId: string, params?: PaginationParams & { status?: string; severity?: string }): Promise<PaginatedResponse<Alert>> {
    return this.client.request<PaginatedResponse<Alert>>('GET', `/smart-cities/cities/${cityId}/alerts`, {
      query: params,
    });
  }

  /**
   * Update alert status
   */
  async updateAlert(alertId: string, status: 'acknowledged' | 'resolved'): Promise<Alert> {
    return this.client.request<Alert>('PATCH', `/smart-cities/alerts/${alertId}`, {
      body: { status },
    });
  }
}
